readme.md

```diff
@@ -1,161 +1,172 @@
+ag /path/to/asyncapi.yaml ./ -o /path/to/generated-code -p moduleName=your-go-module-name
+```
 
-The template provides switches to allow for different levels of generation. More information can be found at [generation levels](#generation-levels)
-* Channel parsing / generation with parameter structs
-* Schema and Message definitions
-* ~~Content Type parsing with the ability to extend supported content types~~ (Soon)
-* A transport interface to support various underlying event messaging protocols
-* ~~A simple wrapping framework to provide AsyncAPI *operations* as high-level functions~~ (Very Soon)
+#### How to use the generated code
 
-## What Am I Getting Myself Into? 
-Code Samples:
+The above code currently generates a golang module that has a AMQP subscriber.
 
-Channel Building / Parsing
-```go
-p := &channel.TurnOnParams{"<my-id>"}
-p.Build() // Returns: "smartylighting/streetlights/1/0/action/<my-id>/turn/on"
+##### Pre-requisites
+To run the generated code the following needs to be installed
 
-p := &channel.ReceiveLightMeasurementParams{}
-err := p.Parse("smartylighting/streetlights/1/0/event/(.+)/lighting/measured")
-// Populates p with channel parameters
-// p = { StreetlightId: "<my-id>" }
-```
+1. go 1.16 +
+2. rabbitmq-server
 
-## Getting Started
-
-### Dependencies:
-The AsyncAPI Generation tool is required to execute this template. Find it here : https://github.com/asyncapi/generator/#install-the-cli.
-
-1. Install AsyncAPI Generation tool
-	```bash
-	$ npm install -g @asyncapi/generator
-	```
-2. Clone the template to wherever you like
-	```bash
-	$ cd ~/dev/tools/
-	$ git clone https://github.com/asyncapi/go-template.git
-	```
-3. Run the generator.  
-	Arguments:
-	* The path to your Async specification file
-	* The path to the template folder of this repo
-	* Destination directory for output
-	```
-	$ ag ./asyncapi.yaml ~/dev/tools/go-template/template -o ./<dest-dir>
-	```
-
-## What does the generation look like? 
-Examples based on the Streetlights example specification
-
-#### Message Sample
-```golang
-//   Location: message/<msg_name/id>.go
-
-
-// DimLight is used to command a particular streetlight to dim the lights.
-type DimLight struct {
-	transport.Message
-
-	// ContentType indicates the specified MIME type for this message. If empty, the defaultContentType should be used 
-	ContentType string
-	
-	// Payload contains the content defined by the payload AsyncAPI field
-	Payload model.DimLightPayload
-}
-```
-#### Schema Sample
-```golang
-//   Location: model/<schema_name>.go
+##### Running the code
 
-// DimLightPayload is a Schema defined in the AsyncAPI specification
-type DimLightPayload struct {
-    // Percentage is a property defined in the AsyncAPI specification
-    Percentage *int64 `json:"percentage"`
+1. Navigate to the path where the code was generated
+2. Run the following commands to download the dependencies
+```bash
+go mod download
+go mod tidy
```

Comment:
if i do not run `go mod tidy` i get an error like

```
missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/subscribers.go:6:5: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill-amqp/pkg/amqp (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
asyncapi/handlers.go:9:2: missing go.sum entry for module providing package github.com/ThreeDotsLabs/watermill/message (imported by go-amqp/asyncapi); to add:
	go get go-amqp/asyncapi
```

Reply:
WDYT about generating the `go.sum` file as well so all checksums are already checked in?

Reply:
since this code is being generated, it would not be an easy task to generate the go.sum
Also from a tooling perspective they way I see it is that this tool is responsible for generating the required code and basic skeleton. From there on it is the responsibility of the consumer to download the required dependencies

Reply:
`go.sum` stores the sum (hash) of all dependencies for consistency check. It is not storing dependencies or anything else. 
Generating the `go.sum` should be an atomic process, so it is as easy as just generating a file containing exactly the same content of your local `go.sum` file.

Then, consumers will need only to do `go mod download` for getting an **exact** copy of what you want.

Reply:
hmm I see what you say, however I have 2 follow up questions

1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?
2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

Reply:
> 1. In future say I want to add a new dependency to the template. Then I will have to add the new dependency in the template and then I would have to generate the code , run `go mod download` so that the `go.sum` file is generated. Is there a good way to automate this process ?

I feel is the natural process to follow. Otherwise, how would you add a dependency to Go generated code without ensuring that generated code can, let's say, compile?
You could perhaps add a end-to-end functional test where you generate code and run testsuite, linters if you want, etc.
I think better to open a new issue in order to track this effort.

> 2. Currently we allow users to pass the `goVersion` as a parameter to this template, will the `go.sum` work across all version of `go`

Whatever version you add as directive to the `go.mod` file indicates the minimum acceptable Go version by that module. Meaning users should not enter a lower version than that.

Reply:
@smoya then should I not provide the `goVersion` option to the generator and say that users must use go 1.16+ in the readme ?

Reply:
`goVersion` option is used **only** to add the constraint to the `go.mod` file. However, that constraint is tied to the generated code, so I think I would +1 to remove it. Maybe if you can validate the `goVersion` option to be 1.16 or greater... but still IMHO is not worth it.

Reply:
this has now been addressed